flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileWriter.java

```diff
@@ -91,176 +94,149 @@ public DataFilePathFactory pathFactory() {
         }
-        return result;
+
+        return rollingKvWriter.result();
     }
 
     public void delete(DataFileMeta file) {
         FileUtils.deleteOrWarn(pathFactory.toPath(file.fileName()));
     }
 
-    private abstract class DataRollingFile extends RollingFile<KeyValue, DataFileMeta> {
+    private class KvBulkWriterFactory implements BulkWriter.Factory<KeyValue> {
+
+        @Override
+        public BulkWriter<KeyValue> create(FSDataOutputStream out) throws IOException {
+            KeyValueSerializer serializer = new KeyValueSerializer(keyType, valueType);
+
+            return new BaseBulkWriter<>(writerFactory.create(out), serializer::toRow);
+        }
+    }
 
+    private class KvFileWriter extends BaseFileWriter<KeyValue, DataFileMeta> {
         private final int level;
-        private final KeyValueSerializer serializer;
         private final RowDataSerializer keySerializer;
 
-        private long rowCount;
-        private BinaryRowData minKey;
-        private RowData maxKey;
-        private long minSequenceNumber;
-        private long maxSequenceNumber;
-
-        private DataRollingFile(int level) {
-            // each level 0 data file is a sorted run,
-            // we must not write rolling files for level 0 data files
-            // otherwise we cannot reduce the number of sorted runs when compacting
-            super(level == 0 ? Long.MAX_VALUE : suggestedFileSize);
+        private FieldStatsCollector keyStatsCollector = null;
+        private FieldStatsCollector valueStatsCollector = null;
+
+        private BinaryRowData minKey = null;
+        private RowData maxKey = null;
+        private long minSeqNumber = Long.MAX_VALUE;
+        private long maxSeqNumber = Long.MIN_VALUE;
+
+        public KvFileWriter(BulkWriter.Factory<KeyValue> writerFactory, Path path, int level)
+                throws IOException {
+            super(writerFactory, path);
+
             this.level = level;
-            this.serializer = new KeyValueSerializer(keyType, valueType);
             this.keySerializer = new RowDataSerializer(keyType);
-            resetMeta();
-        }
-
-        @Override
-        protected Path newPath() {
-            return pathFactory.newPath();
+            if (fileStatsExtractor == null) {
+                this.keyStatsCollector = new FieldStatsCollector(keyType);
+                this.valueStatsCollector = new FieldStatsCollector(valueType);
+            }
         }
 
         @Override
-        protected BulkWriter<RowData> newWriter(FSDataOutputStream out) throws IOException {
-            return writerFactory.create(out);
-        }
+        public void write(KeyValue kv) throws IOException {
+            super.write(kv);
 
-        @Override
-        protected RowData toRowData(KeyValue kv) {
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Writing key-value to data file, kv: " + kv.toString(keyType, valueType));
+            if (fileStatsExtractor == null) {
```

Comment:
Lots of `fileStatsExtractor == null` looks bad.
I think we can have a `StatsProducer` to unify `StatsExtractor` and `StatsCollector`.  To reduce caller complexity. I create https://issues.apache.org/jira/browse/FLINK-27543 for this.